server/v2/appmanager/appmanager.go

```diff
@@ -30,30 +34,50 @@ func (a AppManager[T]) InitGenesis(
 	headerInfo header.Info,
 	consensusMessages []transaction.Type,
 	initGenesisJSON []byte,
+	txDecoder transaction.Codec[T],
 ) (corestore.WriterMap, error) {
+	v, zeroState, err := a.db.StateLatest()
+	if err != nil {
+		return nil, fmt.Errorf("unable to get latest state: %w", err)
+	}
+	if v != 0 {
+		return nil, fmt.Errorf("cannot init genesis on non-zero state")
+	}
+
+	var genTxs []T
+	zeroState, err = a.stf.RunWithCtx(ctx, zeroState, func(ctx context.Context) error {
+		return a.initGenesis(ctx, bytes.NewBuffer(initGenesisJSON), func(jsonTx json.RawMessage) error {
+			genTx, err := txDecoder.DecodeJSON(jsonTx)
+			if err != nil {
+				return fmt.Errorf("failed to decode genesis transaction: %w", err)
+			}
+			genTxs = append(genTxs, genTx)
+			return nil
+		})
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to import genesis state: %w", err)
+	}
 	// run block 0
 	// TODO: in an ideal world, genesis state is simply an initial state being applied
 	// unaware of what that state means in relation to every other, so here we can
 	// chain genesis
 	block0 := &appmanager.BlockRequest[T]{
-		Height:            uint64(headerInfo.Height),
+		Height:            uint64(0),
 		Time:              headerInfo.Time,
 		Hash:              headerInfo.Hash,
 		ChainId:           headerInfo.ChainID,
 		AppHash:           headerInfo.AppHash,
-		Txs:               nil,
+		Txs:               genTxs,
 		ConsensusMessages: consensusMessages,
 	}
 
-	_, genesisState, err := a.DeliverBlock(ctx, block0)
+	_, genesisState, err := a.stf.DeliverBlock(ctx, block0, zeroState)
```

Comment:
```suggestion
	blockResponse, genesisState, err := a.stf.DeliverBlock(ctx, block0, zeroState)
```

returning this all the way down to consensus avoids the need for the query to get the valset